pkg/dbpool/pool.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2023 EMQ Technologies Co., Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/xo/dburl"
+
+	"github.com/lf-edge/ekuiper/v2/internal/conf"
+	"github.com/lf-edge/ekuiper/v2/pkg/dbpool/sqldatabase/driver"
+	"github.com/lf-edge/ekuiper/v2/pkg/hidden"
+)
+
+var GlobalPool *dbPool
+
+func init() {
+	// GlobalPool maintained the *sql.DB group by the driver and DSN.
+	// Multiple sql sources/sinks can directly fetch the `*sql.DB` from the GlobalPool and return it back when they don't need it anymore.
+	// As multiple sql sources/sinks share the same `*sql.DB`, we can directly control the total count of connections by using `SetMaxOpenConns`
+	GlobalPool = newDBPool()
+}
+
+type dbPool struct {
+	sync.RWMutex
+	// url -> *sql.DB
+	pool map[string]*sql.DB
+	// url -> connection count
+	connections map[string]int
+}
+
+func newDBPool() *dbPool {
+	return &dbPool{
+		pool:        map[string]*sql.DB{},
+		connections: map[string]int{},
+	}
+}
+
+func (dp *dbPool) getDBConnCount(url string) int {
+	dp.RLock()
+	defer dp.RUnlock()
+	count, ok := dp.connections[url]
+	if ok {
+		return count
+	}
+	return 0
+}
+
+func (dp *dbPool) getOrCreate(url string) (*sql.DB, error) {
+	hiddenURL, _ := hidden.HiddenURLPasswd(url)
+	dp.Lock()
+	defer dp.Unlock()
+	db, ok := dp.pool[url]
```

Comment:
It looks like the same logic is generalized and implemented in the connection pool. Can we merge this into that?